skema/isa/lib.py

```diff
@@ -0,0 +1,279 @@
+# -*- coding: utf-8 -*
+"""
+All the functions required by performing incremental structure alignment (ISA)
+Author: Liang Zhang (liangzh@arizona.edu)
+Updated date: March 3, 2023
+"""
+
+import warnings
+
+from typing import List, Tuple, Any, Union
+
+from numpy import ndarray
+from pydot import Dot
+
+warnings.filterwarnings('ignore')
+import requests
+import pydot
+import numpy as np
+from graspologic.match import graph_match
+from graphviz import Source
+import graphviz
+from copy import deepcopy
+
+# Set up the random seed
+np.random.seed(1)
+rng = np.random.default_rng(1)
+
+# The encodings of basic operators when converting adjacency matrix
+op_dict = {"+": 1, "-": 2, "*": 3, "/": 4, "=": 5, "√": 6}
+
+'''
+Call the REST API of math-exp-graph to convert the MathML input to its GraphViz representation
+Ensure running the REST API before calling this function
+Input: file directory or the MathML string
+Output: the GraphViz representation (pydot.Dot)
+'''
+
+
+def generate_graph(file: str = "", render: bool = False) -> pydot.Dot:
+    if '<math>' in file and '</math>' in file:
+        content = file
+    else:
+        with open(file) as f:
+            content = f.read()
+
+    digraph = requests.put('http://localhost:8080/mathml/math-exp-graph', data=content.encode('utf-8'))
+    if render:
+        src = Source(digraph.text)
+        src.render('doctest-output/mathml_exp_tree', view=True)
+    graph = pydot.graph_from_dot_data(str(digraph.text))[0]
+    return graph
+
+
+'''
+Convert the GraphViz representation to its corresponding adjacency matrix
+Input: the GraphViz representation
+Output: the adjacency matrix and the list of the names of variables and terms appeared in the expression  
+'''
+
+
+def generate_amatrix(graph: pydot.Dot) -> Tuple[ndarray, List[str]]:
+    node_labels = []
+    for node in graph.get_nodes():
+        node_labels.append(node.obj_dict['attributes']['label'].replace('"', ''))
+
+    amatrix = np.zeros((len(node_labels), len(node_labels)))
+
+    for edge in graph.get_edges():
+        x, y = edge.obj_dict['points']
+        label = edge.obj_dict['attributes']['label'].replace('"', '')
+        amatrix[int(x)][int(y)] = op_dict[label] if label in op_dict else 7
+
+    return amatrix, node_labels
+
+
+'''
```

Comment:
Can you move the docstrings inside the functions? Also, can you run the Python code through the Black autoformatter (https://black.readthedocs.io/en/stable/)?

Reply:
Thanks for letting me know! I have pushed the changes.